core/message-handling.go

```diff
@@ -417,29 +417,38 @@ func makeUIMessageProcessor(captureUI uiCapturer, processViewMessage viewMessage
 	}
 }
 
-func makeViewMessageProcessor(waitView viewWaiter, applyReplicaMessage replicaMessageApplier) viewMessageProcessor {
+func makeViewMessageProcessor(viewState viewstate.State, applyReplicaMessage replicaMessageApplier) viewMessageProcessor {
 	return func(msg messages.ReplicaMessage) (new bool, err error) {
+		var active bool
+
 		switch msg := msg.(type) {
 		case messages.Prepare, messages.Commit:
-			var view uint64
+			var messageView uint64
 
 			switch msg := msg.(type) {
 			case messages.Prepare:
-				view = msg.View()
+				messageView = msg.View()
 			case messages.Commit:
-				view = msg.Prepare().View()
+				messageView = msg.Prepare().View()
 			}
 
-			ok, release := waitView(view)
-			if !ok {
+			currentView, expectedView, release := viewState.HoldView()
+			defer release()
+
+			if currentView == expectedView {
+				active = true
+			}
+
+			if messageView < currentView {
 				return false, nil
+			} else if messageView > currentView {
+				return false, fmt.Errorf("Message refers to unexpected view")
```

Comment:
Good point. If the message refers to an old view then it is simply considered as delayed due to asynchrony in communication; so it gets ignored. However, if the message refers to a future view then the sender is considered faulty, and the error is reported. The latter condition is based on two arguments. 

First, any correct replica, when it starts participating in a new view, must make sure that all other correct replicas will do the same, as described in commit message of "core: Do not wait for active view". 

Second, we use reliable FIFO channels for communication and handle messages received from the same channel sequentially. This implies that messages are processed in the same order as they are sent by the peer. 

Thus, putting it all together, when a replica processes a normal-case message from a correct replica, it must have already transitioned into the corresponding view (or a later view), because the sender must have ensured that before sending the message.

Reply:
I am thinking of adding a comment before returning the error, like: "Correct peer replicas ensure that other replicas transition into a new view before sending them any normal-case message for that view."

Reply:
> First, any correct replica, when it starts participating in a new view, must make sure that all other correct replicas will do the same, as described in commit message of "core: Do not wait for active view".
> 
> Second, we use reliable FIFO channels for communication and handle messages received from the same channel sequentially. This implies that messages are processed in the same order as they are sent by the peer.

I've overlooked that message order is guaranteed to be preserved by gRPC layer. I agree that the 2 points explain the reason well. Thank you.

> I am thinking of adding a comment before returning the error, like: "Correct peer replicas ensure that other replicas transition into a new view before sending them any normal-case message for that view."

Sounds nice.

Reply:
> I've overlooked that message order is guaranteed to be preserved by gRPC layer.

This is also an [API requirement](https://godoc.org/github.com/hyperledger-labs/minbft/api#ReplicaConnector): "The connection guarantees authenticated destination, eventual delivery, preserving integrity and the original order of messages."